core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1000,6 +1034,54 @@ class ReplicaManager(val config: KafkaConfig,
     result
   }
 
+  /**
+    * Using the configured [[ReplicaSelector]], determine the preferred read replica for a partition given the
+    * client metadata, the requested offset, and the current set of replicas
+    */
+  def findPreferredReadReplica(tp: TopicPartition,
+                               clientMetadata: ClientMetadata,
+                               replicaId: Int,
+                               fetchOffset: Long): Option[Int] = {
+    val partition = getPartitionOrException(tp, expectLeader = false)
+
+    if (partition.isLeader) {
+      if (Request.isValidBrokerId(replicaId)) {
+        // Don't look up preferred for follower fetches via normal replication
+        Option.empty
+      } else {
+        val replicaEndpoints = metadataCache.getPartitionReplicaEndpoints(tp.topic(), tp.partition(), new ListenerName(clientMetadata.listenerName))
+        val now = time.milliseconds
+        val replicaInfoSet: mutable.Set[ReplicaView] = partition.allReplicaIds.flatMap(partition.getReplica)
+          // Exclude replicas that don't have the requested offset (whether or not if they're in the ISR)
+          .filter(replica => replica.logEndOffset >= fetchOffset)
+          .filter(replica => replica.logStartOffset <= fetchOffset)
+          .map(replica => new DefaultReplicaView(
+            replicaEndpoints.getOrElse(replica.brokerId, Node.noNode()),
+            replica.logEndOffset,
+            now - replica.lastCaughtUpTimeMs
+          ))
+
+        if (partition.leaderReplicaIdOpt.isDefined) {
+          val leaderReplica: ReplicaView = partition.leaderReplicaIdOpt
+            .map(replicaId => replicaEndpoints.getOrElse(replicaId, Node.noNode()))
+            .map(leaderNode => new DefaultReplicaView(leaderNode, partition.localLogOrException.logEndOffset, 0L))
+            .get
+
+          replicaInfoSet.add(leaderReplica)
```

Comment:
It's actually not in there since `Partition#getReplica` returns None for the leader replica after a recent change. I'll update the code to be more clear.